### DISCUSSION TITLE
Question - TSN: Accessing the input port number of the Switch

### DISCUSSION THREAD

**Discussion by Kamalakannan23** [CoRE-RG/CoRE4INET#43] [Q&A]
Hi,

How to identify the input port number of a TSN flow in the egress port?

In the MACRelayTable there is a function to get the inputport number which is used to update the MAC address table. 

int inputport = frame->getArrivalGate()->getIndex();

I want to do something similar to determine the input port number of a flow in the egress port queues (the 8 egress queues)
I want to access the port number from the PHY port module.

![Switch-MacTable](https://user-images.githubusercontent.com/61795091/181100411-519754fa-dfe3-4877-a623-93c31e559272.PNG)

**Comment by SaloM0nk** [comment]
Hello, 
I don't think that there is an easy way of accessing the input port number at the egress port, since the arrival gate will change when the message is passed from module to module. 

Is there any way that what you want to implement in the output queue can be implemented in the relay engine instead?

If there is no other way you could attach a custom control information to the frame in the switching module (which knows the input port) and than extract that control info again at the egress queue. 

Regards
Timo

**Comment by Kamalakannan23** [comment]
Hello Timo,

Thanks for the information. I have further questions,

_I don't think that there is an easy way of accessing the input port number at the egress port, since the arrival gate will change when the message is passed from module to module._
- Ok.

_Is there any way that what you want to implement in the output queue can be implemented in the relay engine instead?_
- Did you meant to say under MacRelayTable, we were able to access the input port number and to implement them here. No. We want to use it in output queue only. 

_If there is no other way you could attach a custom control information to the frame in the switching module (which knows the input port) and than extract that control info again at the egress queue._
- This seems to be little more difficult (I will try it out)

Whether it is possible to access the mac table using getsubmodule from the egress port?

Any other way that you can suggest to access the input port number from the PHY port module.

Best Regards,
Kamalakannan

**Comment by SaloM0nk** [comment]
Hi Kamalkannan,

> Did you meant to say under MacRelayTable, we were able to access the input port number and to implement them here. No. We want to use it in output queue only.

No I mean the actual relay unit which decides where to forward packets (called beswitch in your image). I thought it might be possible to do decisions based on the ingress port before pushing the packet to an egress port. If not the ControlInformation is the only good way I can think of. 

> Whether it is possible to access the mac table using getsubmodule from the egress port?

Yes it is possible to access the macTable module through a combination of (multiple) getParentModule() calls depending on how deep your Module is in the PHY Port and then calling getSubmodule("macTable ").

Regards
Timo
